blobfile/_common.py

```diff
@@ -648,27 +652,114 @@ def execute_request(
     assert False, "unreachable"
 
 
+class TupleEncoder(json.JSONEncoder):
+    def encode(self, obj):
+        def hint_tuples(item):
+            print(item)
+            if isinstance(item, tuple):
+                return {"__tuple__": True, "items": item}
+            if isinstance(item, list):
+                return [hint_tuples(e) for e in item]
+            if isinstance(item, dict):
+                return {key: hint_tuples(value) for key, value in item.items()}
+            else:
+                return item
+
+        return super(TupleEncoder, self).encode(hint_tuples(obj))
+
+
+def hinted_tuple_hook(obj):
+    if "__tuple__" in obj:
+        return tuple(obj["items"])
+    else:
+        return obj
+
+
 class TokenManager:
     """
     Automatically refresh tokens when they expire
     """
 
     def __init__(
-        self, get_token_fn: Callable[[Config, Any], Tuple[Any, float]]
+        self,
+        get_token_fn: Callable[[Config, Any], Tuple[Any, float]],
+        save_to_disk: bool = True,
     ) -> None:
         self._get_token_fn = get_token_fn
         self._tokens = {}
         self._expirations = {}
         self._lock = threading.Lock()
+        self._save_to_disk = save_to_disk
+
+        if save_to_disk:
+            self._load_token_file()
+
+    def _load_token_file(self):
+        access_tokens_path = os.path.expanduser(ACCESS_TOKEN_FILE)
+        if os.path.exists(access_tokens_path):
+            with open(access_tokens_path) as f:
+                tokens = json.load(f, object_hook=hinted_tuple_hook)
+                for key, value in zip(tokens["token_keys"], tokens["token_values"]):
+                    self._tokens[key] = value
+                for key, value in zip(
+                    tokens["expiration_keys"], tokens["expiration_values"]
+                ):
+                    self._expirations[key] = value
+
+    def _save_token_file(self):
+        access_token_lock_path = os.path.expanduser(ACCESS_TOKEN_LOCK_FILE)
+        access_tokens_path = os.path.expanduser(ACCESS_TOKEN_FILE)
+
+        if not os.path.exists(access_token_lock_path):
+            os.makedirs(os.path.dirname(access_token_lock_path), exist_ok=True)
+            with open(access_token_lock_path, "w") as f:
+                f.write("")
+        lock = filelock.FileLock(access_token_lock_path, timeout=1)
+
+        try:
+            with lock:
+                if not os.path.exists(access_tokens_path):
+                    os.makedirs(os.path.dirname(access_tokens_path), exist_ok=True)
+                with open(access_tokens_path, "w") as f:

```

Comment:
I think you might want to open a temp file and do `os.replace` to atomically replace the existing file.  This isn't fullproof but seems like it would mostly avoid other processes reading truncated or partial file contents.

Reply:
makes sense. Done